fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/cursors/RecursiveUnionCursor.java

```diff
@@ -0,0 +1,264 @@
+            // state.
+            recursiveCursor = recursiveCursorSupplier.get();
+            return onNextRecursive();
+        }
+        return CompletableFuture.completedFuture(RecordCursorResult.exhausted());
+    }
+
+    @Nonnull
+    private CompletableFuture<RecordCursorResult<T>> wrapLastResult(@Nonnull RecordCursorResult<T> innerCursorResult) {
+        return CompletableFuture.completedFuture(RecordCursorResult.withoutNextValue(
+                new Continuation(isInitialState, isReadingFromInitialCursorSupplier,
+                        innerCursorResult::getContinuation, () -> RecordCursorStartContinuation.START, false),
+                innerCursorResult.getNoNextReason()));
+    }
+
+    @Nonnull
+    private CompletableFuture<RecordCursorResult<T>> wrapNextResult(@Nonnull RecordCursorResult<T> innerCursorResult) {
+        final var continuation = new Continuation(isInitialState, isReadingFromInitialCursorSupplier,
+                innerCursorResult::getContinuation, () -> RecordCursorStartContinuation.START, false);
+        return CompletableFuture.completedFuture(RecordCursorResult.withNextValue(innerCursorResult.get(), continuation));
+    }
+
+    @Override
+    public void close() {
+        initialCursor.close();
+        recursiveCursor.close();
+    }
+
+    @Override
+    public boolean isClosed() {
+        return recursiveCursor.isClosed(); // todo: improve.
+    }
+
+    @Nonnull
+    @Override
+    public Executor getExecutor() {
+        return executor;
+    }
+
+    @Override
+    public boolean accept(@Nonnull final RecordCursorVisitor visitor) {
+        if (visitor.visitEnter(this)) {
+            initialCursor.accept(visitor);
+            recursiveCursor.accept(visitor);
+        }
+        return visitor.visitLeave(this);
+    }
+
+    @Nonnull
+    public static <T> RecursiveUnionCursor<T> from(@Nullable final byte[] unparsed,
+                                                   @Nonnull final Function<ByteString, RecordCursor<T>> initialCursorCreator,
+                                                   @Nonnull final Function<ByteString, RecordCursor<T>> recursiveCursorCreator,
+                                                   @Nonnull final Supplier<Boolean> isReadingFromInitialCursorSupplier,
+                                                   @Nonnull final Consumer<Boolean> wasInInitialStateConsumer,
+                                                   @Nonnull final Consumer<Boolean> wasReadingFromFirstTempTableConsumer,
+                                                   @Nonnull final Supplier<Boolean> recursiveStepCompletionCallback,
+                                                   @Nonnull Executor executor) {
+        if (unparsed == null) {
+            final var initialCursor = initialCursorCreator.apply(null);
+            final var recursiveCursor = recursiveCursorCreator.apply(null);
```

Comment:
When is this used, rather than replaced? How can it even be valid, if `initialCursor` was empty, for example?

Reply:
This is now removed, only one cursor is iterated over internally and when it is exhausted (after iterating over all the items from recursive step `n`, the recursive union cursor will either halt, or, if feasible, it will reset that cursor to point to items of recursive step `n+1`.